### ISSUE TITLE
Trying to create a MSE loss function for tfe.keras

### ISSUE THREAD

**Issue by burhanusman** [tf-encrypted/tf-encrypted#654]
Hi,
I'm trying to create MSE loss function similar to the `BinaryCrossentropy` defined in keras.losses.py. 
Can someone look into it and tell me if I'm doing right? 
I'm confused about doing subtraction of a normal tensor with a pondPrivateTensor. 

```
class MSE(Loss):
    """Computes the MSE loss between true
    labels and predicted labels.
    """
    def __init__(self):
      super(MSE, self).__init__(
          mse)

    def grad(self, y_true, y_pred):
      return y_pred - y_true

  def mse(y_true, y_pred):

    batch_size = y_true.shape.as_list()[0]
    batch_size_inv = 1 / batch_size
    out = y_true -y_pred
    out =out.square() #squared error
    mse_loss = out.reduce_sum(axis=0) * batch_size_inv # mean squared error
    #rmse_loss = rmse_loss #root mean square  # TODO
    return mse_loss
```

**Comment by mortendahl** [comment]
> I'm trying to create MSE loss function

Amazing!

> I'm confused about doing subtraction of a normal tensor with a pondPrivateTensor.

What is causing the confusion?

**Comment by burhanusman** [comment]
While training, my loss is not decreasing. I'm worried if I made a mistake in writing the loss function. 
The loss is often negative, which shouldn't be the case since we are squaring the individual terms.

**Comment by mortendahl** [comment]
do you have a runnable example?

**Comment by burhanusman** [comment]
[Black Friday - TFE training - regression .zip](https://github.com/tf-encrypted/tf-encrypted/files/3575074/Black.Friday.-.TFE.training.-.regression.zip)


Here is the full notebook that I was working on. 
What I'm concerned is that I do `out.square` and still my loss values are often negative.

**Comment by mortendahl** [comment]
Hi @burhanusman,

I believe the error is caused by overflow, ie some values become bigger than the (fixedpoint) encoding allows. When that happens the results are undefined, including potentially being negative.

Your `y_...` values do not seem to be normalized, with a max value of 23961. Doing that prevents the overflow (although the network doesn't seem to learn a lot).

**Comment by burhanusman** [comment]
Hey @mortendahl,
That did help! Thanks a lot.
And I see that the network is not learning much. Probably something wrong with the loss function.

**Comment by mortendahl** [comment]
How does your particular network behave on the given data set in plaintext?

**Comment by yanndupis** [comment]
Hey @burhanusman,

Thanks for raising this issue. Just quick head-up I wanted to let you know that I am currently investigating it. 

I ran your notebook with the model in plaintext (normal TF Keras). The model is not learning. I am not familiar with this dataset but there might be some problem with data pre-processing (e.g. Occupation might be a categorical var instead of a numeric one) or the model. 

I am currently experimenting with a mock dataset for linear regression to validate that your approach is working (it should :)). However I think I might have ran in a bug. I will keep you updated. 

Thank you!

**Comment by burhanusman** [comment]
Thanks a lot @yanndupis for looking into this.
As Mortendahl mentioned, the overflow issue exist if y is not normalised. I did try normalising it, and though the loss is no more negative, the model isn't learning much. Also the model is not learning a lot in plaintext.

I used a model with one more layer in plaintext and it seems like it's learning now. This is the model i used. 
```
# Define model
model1 = Sequential()
model1.add(Dense(100, batch_input_shape=(80,11), activation= "relu"))
model1.add(Dense(50, activation= "relu"))
model1.add(Dense(1))
```

But when I'm using the above model architecture with TFE, it's giving negative loss, even with normalised y. I'm currently investigating it.
```
model = tfe.keras.Sequential()
model.add(tfe.keras.layers.Dense(100, batch_input_shape=[batch_size, num_features]))
model.add(tfe.keras.layers.Activation('sigmoid'))
model.add(tfe.keras.layers.Dense(50))
model.add(tfe.keras.layers.Activation('sigmoid'))
model.add(tfe.keras.layers.Dense(1))
```

Also, @yanndupis , I hope you looked at my definition of MSE loss. Is that the right way to go about it?

Thanks a lot!

**Comment by yanndupis** [comment]
Ok great @burhanusman! no problem. 

For the MSE loss I think the forward pass is correct (`mse()`) however for the gradient I think it should be: 

```
def grad(self, y_true, y_pred):
    batch_size = y_true.shape.as_list()[0] 
    batch_size_inv = 1 / batch_size  
    return 2 * (y_true - y_pred) * batch_size_inv 
```

Just a tip which help me to debug encrypted ML (maybe you are already doing it), is to use a very basic dataset (you can generate a mock dataset) with the simplest  model (just a dense layer). 

You can find a [notebook here](https://github.com/tf-encrypted/tf-encrypted/blob/mse_debug/examples/notebooks/keras-training/Keras_linear_regression.ipynb) with an example (plaintext and tfe implementation). The loss is not exploding however the loss is increasing in this implementation so there is another bug somewhere :) 

I will continue to check this tomorrow but do not hesitate to play with this notebook if you want. 

Thanks!

**Comment by yanndupis** [comment]
ah my bad, in the grad, I inverted y_true and y_pred:

```
def grad(self, y_true, y_pred):
    batch_size = y_true.shape.as_list()[0] 
    batch_size_inv = 1 / batch_size  
    return 2 * (y_pred - y_true) * batch_size_inv 
```

It's training correctly in the [notebook example](https://github.com/tf-encrypted/tf-encrypted/blob/mse_debug/examples/notebooks/keras-training/Keras_linear_regression.ipynb). 

If you could submit a PR with MSE loss (including test) and validate it's behaving correctly for your dataset/model, that would be fantastic. 

Thank you!

**Comment by mortendahl** [comment]
> If you could submit a PR with MSE loss (including test) and validate it's behaving correctly for your dataset/model, that would be fantastic.

hi @burhanusman, just checking in to see if the above is something you'd be up for? if not we'd be happy to get your code into a commit and give you credit in the description.

**Comment by burhanusman** [comment]
Hey @mortendahl ,
This is something I would really love to do and do my bit for this wonderful project. 
I was a bit caught up with other stuffs. Give me day or two if that's all right, and I'll submit a PR. 
Thanks.

**Comment by mortendahl** [comment]
No worries, take your time! Very happy to hear you'd like to contribute.

**Comment by burhanusman** [comment]
It's just that it yet doesn't work for my dataset. 
Both the plaintext and encrypted networks doesn't learn when I normalize the target values.
Plaintext network learns fine with the actual target values, but if I do that for encrypted network, the overflow happens and I get -ve losses. I'm figuring out how to go about that.

**Comment by mortendahl** [comment]
what does the distribution of the target values look like after normalization?

@yanndupis might have some ideas if you're looking..?

**Comment by burhanusman** [comment]
Hey @mortendahl,
Here is the the distribution of the target values after normalization.
![image](https://user-images.githubusercontent.com/7271865/65573246-f7a6e600-df87-11e9-802b-955800410bc9.png)'

I run the same model on actual and normalized target values and I'm getting drastically different results for same epochs of training. 

I know the issue has nothing to do with TF-encrypted now. But thanks for taking time.

**Comment by burhanusman** [comment]
Thannks @mortendahl and @yanndupis .
It's working now. I'll submit a PR soon.
